// @typespec/compiler/dist/init/init-template.js
import { TypeSpecConfigJsonSchema } from "../config/config-schema.js";
export const InitTemplateSchema = {
    type: "object",
    additionalProperties: false,
    properties: {
        title: { type: "string" },
        description: { type: "string" },
        libraries: { type: "array", items: { type: "string" } },
        skipCompilerPackage: { type: "boolean", nullable: true },
        config: { nullable: true, ...TypeSpecConfigJsonSchema },
        inputs: {
            type: "object",
            nullable: true,
            additionalProperties: {
                type: "object",
                properties: {
                    description: { type: "string" },
                    type: { type: "string", enum: ["text"] },
                    initialValue: {},
                },
                required: ["description", "type"],
            },
            required: [],
        },
        files: {
            type: "array",
            nullable: true,
            items: {
                type: "object",
                properties: {
                    path: { type: "string" },
                    destination: { type: "string" },
                },
                required: ["path", "destination"],
            },
        },
    },
    required: ["title", "description"],
};
export const InitTemplateDefinitionsSchema = {
    type: "object",
    additionalProperties: InitTemplateSchema,
    required: [],
};
